uamqp/client.py

```diff
@@ -640,13 +640,24 @@ def receive_message_batch(self, max_batch_size=None, on_message_received=None, t
             return batch
 
         while receiving and not expired and len(batch) < max_batch_size:
-            receiving = self.do_work(batch=max_batch_size)
+
+            while receiving and self._received_messages.qsize() < max_batch_size:
+                if timeout and time.time() > timeout:
+                    expired = True
+                    break
+
+                before = self._received_messages.qsize()
+                receiving = self.do_work(batch=max_batch_size)
+                received = self._received_messages.qsize() - before
```

Comment:
I wonder if it would be more efficient to combine the "did we receive anything" logic with the "get_nowait" logic below - so we only need to count/pull from the queue once. Would that work?

Reply:
I think it would work, I have updated the algorithm here to use counter instead of relying on queue/array directly.
let me know it looks you